website/pages/blog/graphql-shared-stateful-subscription-with-rxjs.mdx

```diff
@@ -0,0 +1,335 @@
+
+Let's modify our publisher factory function:
+```typescript
+const makePublisher = (entityId: number): Observable<number> => {
+  return interval(1000).pipe(
+    // clean up publisher cache when no more subscribers to this publisher
+    finalize(() => publishersMap.delete(entityId)),
+    share(),
+  );
+};
+```
+
+### Delay the shared observable cleanup
+
+Because the view might be expensive to fetch and update, we only do so when users are subscribing to it. In the previous example, as soon as all users have unsubscribed, the shared observable is unsubscribed and deleted from the cache. This might not be the desired behavior, e.g.:
+- we can have one user subscribing to the view, this user refreshes his browser's page, effectively ending the subscription, triggering a cache cleanup, but then the view is resubscribed almost immediately, triggering a new fresh fetch
+- we can have one user stopping the subscription (e.g. navigating away from this view to another part of our app), but another user subscribing soon (a few milliseconds, seconds or minutes) after
+
+To optimize for those use cases, we can delay the unsubscription of the shared observable when no more subscriber are subscribing to it, using the `resetOnRefCountZero` config of the `share` operator. This is available in RXJS v7.
+
+```typescript
+const makePublisher = (entityId: number): Observable<number> => {
+  return interval(1000).pipe(
+    // clean up publisher cache when no more subscribers to this publisher
+    finalize(() => publishersMap.delete(entityId)),
+    share({
+      // when no more subscribers, wait 10 seconds to unsubscribe from the source and reset the underlying subject
+      resetOnRefCountZero: () => timer(10_000),
+    }),
+  );
+};
+```
+
+In the example above, after all subscribers have unsubscribed (`refcount` is 0) the connector (the subject that subscribes to the source observable) now waits for 10 seconds (10000 ms) before it resets the shared observable. During this period of time, the observable pipeline continues to execute. If a new subscriber subscribes before the `timer` has emitted, the `refcount` is again greater than 0 and the connector will keep connecting to the source observable.
+
+<Callout type="warning">
+  We use `timer` in this example but we can use any observable (e.g. receiving a specific event from our message queue) that emits at some point. Be careful though to make sure that the observable you use will actually emit otherwise you'll have a memory leak. To remedy this issue, you might want to add a safety, e.g. using [`raceWith`](https://rxjs.dev/api/operators/raceWith) with `timer` to make sure you don't wait indefinitely:
```

Comment:
```suggestion
  We use `timer` in this example but we can use any observable (e.g. receiving a specific event from our message queue) that emits at some point. Be careful though to make sure that the observable you use will actually emit otherwise you'll have a memory leak. To remedy this issue, you might want to add safety, e.g. using [`raceWith`](https://rxjs.dev/api/operators/raceWith) with `timer` to make sure you don't wait indefinitely:
```